source/common/config/grpc_mux_impl.cc

```diff
@@ -94,37 +94,32 @@ GrpcMuxWatchPtr GrpcMuxImpl::addWatch(const std::string& type_url,
   return watch;
 }
 
-void GrpcMuxImpl::pause(const std::string& type_url) {
-  ENVOY_LOG(debug, "Pausing discovery requests for {}", type_url);
-  ApiState& api_state = api_state_[type_url];
-  ASSERT(!api_state.paused_);
-  ASSERT(!api_state.pending_);
-  api_state.paused_ = true;
+ScopedResume GrpcMuxImpl::pause(const std::string& type_url) {
+  return pause(std::vector<std::string>{type_url});
 }
 
-void GrpcMuxImpl::pause(const std::vector<std::string> type_urls) {
+ScopedResume GrpcMuxImpl::pause(const std::vector<std::string> type_urls) {
   for (const auto& type_url : type_urls) {
-    pause(type_url);
-  }
-}
-
-void GrpcMuxImpl::resume(const std::string& type_url) {
-  ENVOY_LOG(debug, "Resuming discovery requests for {}", type_url);
-  ApiState& api_state = api_state_[type_url];
-  ASSERT(api_state.paused_);
-  api_state.paused_ = false;
-
-  if (api_state.pending_) {
-    ASSERT(api_state.subscribed_);
-    queueDiscoveryRequest(type_url);
-    api_state.pending_ = false;
+    ENVOY_LOG(debug, "Pausing discovery requests for {}", type_url);
+    ApiState& api_state = api_state_[type_url];
+    ASSERT(!api_state.paused_);
+    ASSERT(!api_state.pending_);
```

Comment:
Do you want to reference count or is that for another PR?

Reply:
talked offline, we want to have clearer trace of pause/resume. In ideal world we should always resume in the scope of the pause callsite. at present, clustermanagerimpl updateClusterCounts (together with cds::onConfigUpdate) is the only place that doesn't meet this 'requirement'.
if we can achieve the pause/resume in a single scope goal, i'd want to keep these asserts. If not, we can remove them in a followup PR.